nvim-wrap.1.in
.TH NVIM-WRAP 1 nvim-wrap\-@VERSION@
.SH NAME
nvim-wrap \- a neovim wrapper to set the RPC pipe socket
.SH SYNOPSIS
.B nvim-wrap
.RB [ "nvim options" ]
.IR file
.PP
.B nvim-send
.RB [ commands ]
.SH DESCRIPTION
The program
.IR nvim-wrap
is a wrapper script to set the neovim RPC pipe socket to a user define-able
location, this facilitates user scripting of the neovim RPC capabilities through the
.IR nvim-send
program which will dispatch the neovim commands to every running neovim instance with a pipe file
registered on the common location.

For information on the neovim program consult
.B nvim
(1)

Internally the wrappers are rather simple.

In the case of
.IR nvim-wrap
just executes the neovim binary with the
.B --listen
flag and sets the name of the pipe to be the nvim.PID.timestamp where
.B PID
is the Process ID of the wrapping shell that will be inherited onto the neovim instance through
exec and
.B timestamp
is the number of seconds since the unix epoch.

Meanwhile
.IR nvim-send
defines the internal nvim_send function that uses the
.B --server
with the pipe dir location defined on the nvim-wrap config and the
.B --remote-send
flag which will be populated by whichever string is passed onto the function and terminated with the
neovim caret return <CR> command line sequence, the nvim_send function takes care of dispatching the
commands to every neovim instance, so all needed is to pass along a string as the argument to
nvim-send, however if nvim-send is not passed an argument the internal function send_comms will be
ran, the function is by default empty and is intended to be user overwriteable inside the config
file.

.SH OPTIONS
.TP
.B "help, \-h, \-\-help"
Print brief usage information to standard output and exit.
.TP
.B "version, \-v, \-\-version"
print version information then exit, for nvim-wrap it will also print neovim's version info.
.SH CONFIGURATION
.P
Even tho the commands nvim-wrap and nvim-send interoperate they have different configuration files,
the format for the configuration file is a shell, meaning that variables and functions can be
defined and modified inside the configuration at the whim of the users to create additional
customizations or workarounds for their specific environment and desired neovim experience.

.P
.IR NVIM-WRAP

The configuration of nvim-wrap is located at:
.B ${XDG_CONFIG_HOME:-${HOME}/.config}/nvim-wrap/configrc

.P
To see the location of the configuration file and generating it if running for the first time, run
nvim-wrap with the help flag

.P
The following options are available for the config file:

.TP
.B
nv_bin
the neovim binary to use, by default
.B /usr/bin/nvim
is used, but this value can be changed to the distribution's reccomended path for the neovim binary,
you use an expansion from a subshell running the
.B command
(1)
command with the -v flag to set the neovim binary dynamically like the following example if you'd
like.

.RS
.EX
.eo
nv_bin="$(command -v nvim)"
.ec
.EE
.RE


.TP
.B
pipe_loc
the directory where the pipe files for every neovim instance launched through nvim-wrap will be
located at
.B ${XDG_RUNTIME_DIR:-/tmp}/nvim
is used, this assignment will try to use the value of the XDG_RUNTIME_DIR environment variable, were
it not set it will default to the system's tmp dir.


.P
.IR NVIM-SEND

The configuration of nvim-send is located at:
.B ${XDG_CONFIG_HOME:-${HOME}/.config}/nvim-send/configrc

.P
To see the location of the configuration file and generating it if running for the first time, run
nvim-send with the help flag

.P
The configuration for nvim-send is somewhat simpler as it does not define any option, for the values
of the
.B nv_bin
and
.B pipe_loc
options nvim-send just outright loads the configuration from nvim-wrap, instead what you get is
again a shell format file that is loaded rather than executed and defines just the empty function
.IR send_comms
which is intended to be user populated, this function will be ran in the case that nvim-send is
executed without any arguments.

.SH AUTHOR
.EX
eylles                 <ed.ylles1997 at gmail.com>
.EE
.SH HOMEPAGE
.TP
Source code:
.EE
https://github.com/eylles/nvim-wrap
.EX


nvim-wrap.sh
#!/bin/sh

myname="${0##*/}"

# you can change this on config to the correct path on your system
nv_bin="/usr/bin/nvim"

pipe_loc="${XDG_RUNTIME_DIR:-/tmp}/nvim"

config_dir="${XDG_CONFIG_HOME:-${HOME}/.config}/nvim-wrap"
config_file="${config_dir}/configrc"

# loading the config here means the user can overwrite any of the functions
if [ -f "$config_file" ]; then
    . "$config_file"
else
    if [ ! -d "$config_dir" ]; then
        mkdir -p "$config_dir"
    fi
    cat << __HEREDOC__ >> "$config_file"
# vim: ft=sh
# ${myname} config file

# neovim binary
nv_bin="${nv_bin}"

# pipe directory
pipe_loc="\${XDG_RUNTIME_DIR:-/tmp}/nvim"
__HEREDOC__
fi

version=@VERSION@

show_help () {
    printf '%s\n\t%s\n' "${myname}:" \
                        "nvim wrapper"
    printf '%s\n\t%s\n' "Version:" "$version"
    printf '%s\n\t%s\n' "Usage:" \
                        "${myname} -v | -h | [ nvim flags and files ]"
    printf '%s\n\t%s\n' "    -h" \
                        "print out this help message, 'help' and '--help' are supported too"
    printf '%s\n\t%s\n' "    -v" \
                        "print out the version, 'version' and '--version' are supported too"
    printf '\n%s\n'     "About:"
    printf '\t%s\n'     "$myname is a neovim wrapper intended to explicitly set the location of the"
    printf '\t%s\n'     "neovim rpc pipe socket in a user configurable location, it also allows for"
    printf '\t%s\n'     "setting, unsetting and otherwise modifiying the environment variables"
    printf '\t%s\n'     "neovim is launched with through the configuration."
    printf '\t%s\n'     "The default location is set on: \${XDG_RUNTIME_DIR:-/tmp}/nvim/"
    printf '\t%s\n'     "altho the configuration for nvim-wrap will be used to determine the dir"
    printf '\t%s\n'     "containing the rpc socket for each neovim instance, which by convention is"
    printf '\t%s\n'     "given the name 'nvim.PID.timestamp' where 'PID' is the process id of each"
    printf '\t%s\n'     "instance and timestamp is the seconds since epoch."
    printf '\n%s\n'     "Configuration:"
    printf '\t%s\n'     "The $myname base configuration is generated upon first run, it will be"
    printf '\t%s'       "located by default at: "
    printf '%s\n'       "\${XDG_CONFIG_HOME:-\${HOME}/.config}/nvim-wrap/configrc"
    printf '\t%s'       "on this system that expands to: "
    printf '%s\n'       "$config_file"
    printf '\n%s\n'     "For more information please check the manual page for nvim-wrap(1)"
}

pipe_from_args=""

ARGSC="$#"

while [ "$ARGSC" -gt 0 ]; do case "$1" in
    help|-h|--help)
        show_help
        $nv_bin --help
        exit 0
        ;;
    version|-v|--version)
        printf "%s\n" "nvim-wrap $version"
        $nv_bin --version
        exit 0
        ;;
    --listen)
        if [ -n "$2" ]; then
            pipe_file="$2"
            pipe_from_args=1
            shift 2  # Remove --listen and its value from the arguments
            ARGSC=$((ARGSC - 2))
        else
            echo "Error: --listen requires a file path argument."
            exit 1
        fi
        ;;
    *)
        # Preserve other arguments
        set -- "$@" "$1"
        shift
        ARGSC=$((ARGSC - 1))
        ;;
esac done

if [ -z "$pipe_from_args" ]; then
    mypid="$$"

    timestamp="$(date '+%s')"

    pipe_file="${pipe_loc}/nvim.${mypid}.${timestamp}"
else
    pipe_loc="${pipe_file%/*}"
fi

# make pipe dir
if [ ! -d "$pipe_loc" ]; then
    mkdir -p "$pipe_loc"
fi

exec $nv_bin --listen "$pipe_file" "$@"
